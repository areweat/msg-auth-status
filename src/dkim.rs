//! Public DKIM types exported by this library

#[derive(Clone, Debug, Default, PartialEq)]
pub struct DkimResult<'hdr> {
    pub code: DkimResultCode,
    pub reason: Option<&'hdr str>,
    pub header_d: Option<&'hdr str>,
    pub header_i: Option<&'hdr str>,
    pub header_b: Option<&'hdr str>,
    pub header_a: Option<DkimAlgorithm<'hdr>>,
    pub header_s: Option<&'hdr str>,
    pub raw: Option<&'hdr str>,
}

impl<'hdr> DkimResult<'hdr> {
    pub(crate) fn set_header(&mut self, prop: &ptypes::DkimHeader<'hdr>) -> bool {
        match prop {
            ptypes::DkimHeader::D(val) => self.header_d = Some(val),
            ptypes::DkimHeader::I(val) => self.header_i = Some(val),
            ptypes::DkimHeader::B(val) => self.header_b = Some(val),
            ptypes::DkimHeader::A(val) => self.header_a = Some(val.clone()),
            ptypes::DkimHeader::S(val) => self.header_s = Some(val),
            _ => {}
        }
        true
    }
}

/// DKIM Result Codes - s.2.7.1
//#[derive(Debug, Default, EnumString, StrumDisplay)]
//#[strum(serialize_all = "lowercase", ascii_case_insensitive)]
#[derive(Clone, Debug, Default, PartialEq)]
pub enum DkimResultCode {
    #[default]
    Unknown,
    /// The message was not signed.
    NoneDkim,
    /// The message was signed, the signature or signatures were
    /// acceptable to the ADMD, and the signature(s) passed verification
    /// tests.
    Pass,
    /// The message was signed and the signature or signatures were acceptable
    /// to the ADMD, but they failed the verification test(s).
    Fail,
    /// The message was signed, but some aspect of the signature or
    /// signatures was not acceptable to the ADMD.
    Policy,
    /// The message was signed, but the signature or signatures
    /// contained syntax errors or were not otherwise able to be
    /// processed.  This result is also used for other failures not
    /// covered elsewhere in this list.
    Neutral,
    /// The message could not be verified due to some error that
    /// is likely transient in nature, such as a temporary inability to
    /// retrieve a public key.  A later attempt may produce a final
    /// result.
    TempError,
    /// The message could not be verified due to some error that
    /// is unrecoverable, such as a required header field being absent.
    /// A later attempt is unlikely to produce a final result.
    PermError,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub enum DKimCanonicalization<'hdr> {
    #[default]
    /// simple/simple & simple algorithm tolerates almost no modification
    Simple,
    /// relaxed/simple & relaxed algorithm tolerates common modifications such
    /// as whitespace replacement and header field line rewrapping.    
    Relaxed,
    /// Unknown RFC does not define
    Unknown(&'hdr str),
}

#[derive(Clone, Debug, Default, PartialEq)]
pub enum DkimTimestamp<'hdr> {
    #[default]
    Unknown,
    Raw(&'hdr str),
}

#[derive(Clone, Debug, PartialEq)]
#[allow(non_camel_case_types)]
pub enum DkimAlgorithm<'hdr> {
    /// Do not use
    Rsa_Sha1,
    /// Widely supported
    Rsa_Sha256,
    /// Please support this - not widely supported yet
    Ed25519_Sha256,
    Unknown(&'hdr str),
}
#[derive(Clone, Debug, PartialEq)]
pub enum DkimVersion<'hdr> {
    /// RFC just says this should be used
    One,
    /// Something else outside RFC
    Unknown(&'hdr str),
}

pub mod ptypes;
pub use ptypes::DkimProperty;
