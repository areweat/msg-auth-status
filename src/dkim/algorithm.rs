//! DKIM Algorithms

/// DKIM Algorithms per IANA and RFC
#[derive(Clone, Debug, PartialEq)]
#[allow(non_camel_case_types)]
pub enum DkimAlgorithm<'hdr> {
    /// Do not use
    Rsa_Sha1,
    /// Widely supported
    Rsa_Sha256,
    /// Please support this - not widely supported yet
    Ed25519_Sha256,
    /// Unknown algorithm not specified by RFC / IANA
    Unknown(&'hdr str),
}

/// Currently no error - may change in future
#[derive(Debug, PartialEq)]
pub enum DkimAlgorithmError {}

impl<'hdr> TryFrom<&'hdr str> for DkimAlgorithm<'hdr> {
    type Error = DkimAlgorithmError;

    fn try_from(algo: &'hdr str) -> Result<Self, Self::Error> {
        let ret = match algo {
            "rsa-sha1" => Self::Rsa_Sha1,
            "rsa-sha256" => Self::Rsa_Sha256,
            "ed25519-sha256" => Self::Ed25519_Sha256,
            _ => Self::Unknown(algo),
        };
        Ok(ret)
    }
}
