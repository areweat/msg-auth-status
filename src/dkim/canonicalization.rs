#[derive(Clone, Debug, Default, PartialEq)]
pub enum DkimCanonicalization<'hdr> {
    #[default]
    /// simple/simple & simple algorithm tolerates almost no modification
    Simple,
    /// relaxed/simple & relaxed algorithm tolerates common modifications such
    /// as whitespace replacement and header field line rewrapping.
    Relaxed,
    /// Unknown RFC does not define
    Unknown(&'hdr str),
}

#[derive(Clone, Debug, PartialEq)]
pub enum DkimCanonicalizationError {}

impl<'hdr> TryFrom<&'hdr str> for DkimCanonicalization<'hdr> {
    type Error = DkimCanonicalizationError;

    fn try_from(hdr: &'hdr str) -> Result<Self, Self::Error> {
        let ret = match hdr {
            "simple" => Self::Simple,
            "simple/simple" => Self::Simple,
            "relaxed" => Self::Relaxed,
            "relaxed/simple" => Self::Relaxed,
            _ => Self::Unknown(hdr),
        };
        Ok(ret)
    }
}
