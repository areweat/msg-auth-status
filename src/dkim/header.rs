//! DKIM header.* or header tags per RFC 6376

use super::*;

/// See RFC 6376 s. 3.5 for the full definitions
#[derive(Clone, Debug, PartialEq)]
pub enum DkimHeader<'hdr> {
    /// Required - Version
    V(DkimVersion<'hdr>),
    /// Signature Algorithm - see s. 3.3 & IANA
    A(DkimAlgorithm<'hdr>),
    /// Signature data in base64 (note about FWS in s. 3.5 b=)
    B(&'hdr str),
    /// Hash of canonicalized body part of the message as limited by the 'l=' Body length limit tag - base64.
    /// Note: Whitespaces / WHS are ignored
    Bh(&'hdr str),
    /// Required -  Message canonicalization informs the verifier of the type of canonicalization used to prepare the message for signing. See s.3.4
    C(DkimCanonicalization<'hdr>),
    /// Required - The SDID claiming responsibility for an introduction of a message into the mail stream.
    /// The SDID MUST correspond to a valid DNS name under which the DKIM key ecord is published.
    D(&'hdr str),
    /// Required - Signed header fields separated by colon ':' - see 'h='
    H(&'hdr str),
    /// Optional - The Agent or User Identifier (AUID) on behalf of which the SDID is taking responsibility.
    I(&'hdr str),
    /// Optional - Body length limit - see misuse on RFC 6376 s. 8.2.
    L(&'hdr str),
    /// Optional - Query method - currently only Dns.
    Q(&'hdr str),
    /// Required - The selector subdividing the namespace for the "d=" (domain) tag.
    /// Internationalized selector names MUST be encoded as A-labels, as described in Section 2.3 of RFC 5890.
    S(&'hdr str),
    /// Recommended - Signature Timestamp
    T(DkimTimestamp<'hdr>),
    /// Recommended - Signature Expiration
    X(DkimTimestamp<'hdr>),
    /// Optional - Copied header fields
    Z(&'hdr str),
    /// RFC 6541
    Atps(&'hdr str),
    /// RFC 6541
    Atpsh(&'hdr str),
    /// RFC 6651
    R(&'hdr str),
    /// RFC 5322
    Rfc5322From(&'hdr str),
    /// Unknown
    Unknown(&'hdr str, &'hdr str),
}

/// Currently no errors - may change in the future
#[derive(Clone, Debug, PartialEq)]
pub enum DkimHeaderError {}
