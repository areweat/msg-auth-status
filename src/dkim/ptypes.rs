//! DKIM Property types or 'ptype' per RFC 8601 s. 2.7.1
//!
//! IANA Maintains the registry for all the possible Parameters
//!
//! DKIM-Signature Tag Specifications are defined in RFC 6376 s. 7.2
//!
//! Also see s. 7.10 for the DKIM-Signature field

use super::*;

/// DKIM ptypes
#[derive(Clone, Debug, PartialEq)]
pub enum DkimProperty<'hdr> {
    /// header.* subset of RFC 6376 tags
    Header(DkimHeader<'hdr>),
    /// policy.*
    Policy(DkimPolicy<'hdr>),
}

/// DKIM policy ptype properties
#[derive(Clone, Debug, PartialEq)]
pub enum DkimPolicy<'hdr> {
    /// Unknown
    Unknown(&'hdr str, &'hdr str),
}
