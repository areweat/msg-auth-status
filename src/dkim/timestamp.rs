#[derive(Clone, Debug, Default, PartialEq)]
pub enum DkimTimestamp<'hdr> {
    #[default]
    Unknown,
    Raw(&'hdr str),
}

#[derive(Debug, PartialEq)]
pub enum DkimTimestampError {}

impl<'hdr> TryFrom<&'hdr str> for DkimTimestamp<'hdr> {
    type Error = DkimTimestampError;

    fn try_from(in_str: &'hdr str) -> Result<Self, Self::Error> {
        Ok(Self::Raw(in_str))
    }
}
