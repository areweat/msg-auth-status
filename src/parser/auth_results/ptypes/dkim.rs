//! Parsing dkim property types & values

use super::*;

pub mod dkim_property_key;
pub mod dkim_property_value;
