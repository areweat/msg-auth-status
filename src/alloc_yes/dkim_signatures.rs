//! Allocating DkimSignaturesHandler

use crate::dkim::DkimSignature;

/// TODO: Errors
#[derive(Debug, PartialEq)]
pub enum DkimSignaturesError {}

/// Allocating parsed results for DKIM-Signatures
#[cfg(any(feature = "alloc", feature = "std"))]
#[derive(Debug, Default)]
pub struct DkimSignatures<'hdr> {
    /// DKIM-Signature results
    dkim_signatures: Vec<DkimSignature<'hdr>>,
}

impl<'hdr> crate::traits::ResultsVerifier for DkimSignatures<'hdr> {
    fn return_path_atleast_one_dkim_pass(&self, _selector: &str) -> bool {
        false // TODO
    }
}

impl<'hdr> DkimSignatures<'hdr> {
    /// Parse all DKIM Signatures into allocating Vec from mail_parser::Message
    #[cfg(feature = "mail_parser")]
    pub fn from_mail_parser(
        msg: &'hdr mail_parser::Message<'hdr>,
    ) -> Result<Self, DkimSignaturesError> {
        let mut new_self = Self {
            dkim_signatures: vec![],
        };

        new_self.dkim_signatures = msg
            .header_values("DKIM-Signature")
            .map(|mh| mh.try_into().unwrap())
            .collect();

        Ok(new_self)
    }
}

#[cfg(test)]
#[cfg(feature = "mail_parser")]
mod test {
    use super::*;
    use insta::assert_debug_snapshot;
    use rstest::rstest;
    use std::{fs::File, io::Read, path::PathBuf};

    fn load_test_data(file_location: &str) -> Vec<u8> {
        let mut file = File::open(file_location).unwrap();
        let mut data: Vec<u8> = vec![];
        file.read_to_end(&mut data).unwrap();
        data
    }

    #[rstest]
    fn from_mail_parser(#[files("test_data/*.eml")] file_path: PathBuf) {
        let new_snapshot_path = file_path.with_extension("snap");

        insta::with_settings!({snapshot_path => new_snapshot_path}, {
            insta::allow_duplicates! {
                let raw  = load_test_data(file_path.to_str().unwrap());
                let parser = mail_parser::MessageParser::default();
                let parsed_message = parser.parse(&raw).unwrap();
                let status = DkimSignatures::from_mail_parser(&parsed_message).unwrap();
                assert_debug_snapshot!(&status);
            }
        });
    }
}
