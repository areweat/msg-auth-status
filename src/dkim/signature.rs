//! DKIM Signatures

use crate::dkim::*;

/// RFC 6376 s. 3.5
#[derive(Clone, Debug, PartialEq)]
pub struct DkimSignature<'hdr> {
    /// Version
    pub v: DkimVersion<'hdr>,
    /// Algorithm
    pub a: DkimAlgorithm<'hdr>,
    /// Signature data (base64)
    pub b: &'hdr str,
    /// Hash of canonicalized body part of the message as limited by the 'l='
    pub bh: &'hdr str,
    /// Message canonicalization informs the verifier of the type of canonicalization used to prepare the message for signing. See s.3.4
    pub c: Option<DkimCanonicalization<'hdr>>,
    /// The SDID claiming responsibility for an introduction of a message into the mail stream
    pub d: &'hdr str,
    /// Signed header fields separated by colon ':' - see 'h='
    pub h: &'hdr str,
    /// The Agent or User Identifier (AUID) on behalf of which the SDID is taking responsibility.
    pub i: Option<&'hdr str>,
    /// Body length limit - see misuse on RFC 6376 s. 8.2
    pub l: Option<&'hdr str>,
    /// Query methods - currently only DnsTxt
    pub q: Option<&'hdr str>,
    /// The selector subdividing the namespace for the "d=" (domain) tag
    pub s: &'hdr str,
    /// Recommended - Signature Timestamp
    pub t: Option<DkimTimestamp<'hdr>>,
    /// Recommended - Signature Expiration
    pub x: Option<DkimTimestamp<'hdr>>,
    /// Copied header fields
    pub z: Option<&'hdr str>,
}

/// DKIM-Signature header parsing Errors
#[derive(Debug)]
pub enum DkimSignatureError {
    /// No tags found at all
    NoTagFound,
    /// Encountered unexpected Equal '=' character when a tag was expected
    UnexpectedEqual,
    /// Error (from lexer) trying to parse value - Unmatched
    ParseValueUnmatch,
    /// Error (from parse conversion) trying to parse a valid value on tag
    ParseValueInvalid(DkimTagValueError),
    /// Missign Required v=
    MissingVersion,
    /// Msising Required a=
    MissingAlgorithm,
    /// Msising Required b=
    MissingSignature,
    /// Missing Required bh=
    MissingBodyHash,
    /// Missing Required d=
    MissingResponsibleSdid,
    /// Msising Required h=
    MissingSignedHeaderFields,
    /// Missing Required s=
    MissingSelector,
}

/// DKIM Header tags values parsing error
#[derive(Debug, PartialEq)]
pub enum DkimTagValueError {
    /// Tag value must appear only once per tag
    Duplicate,
    /// DKIM Timestamp parsing error
    Timestamp(DkimTimestampError),
    /// DKIM Canonizalition parsing error
    Canonicalization(DkimCanonicalizationError),
    /// DKIM Algorithm parsing error
    Algorithm(DkimAlgorithmError),
    /// DKIM Version parsing error
    Version(DkimVersionError),
}

impl From<DkimTagValueError> for DkimSignatureError {
    fn from(e: DkimTagValueError) -> Self {
        Self::ParseValueInvalid(e)
    }
}

impl From<DkimTimestampError> for DkimTagValueError {
    fn from(e: DkimTimestampError) -> Self {
        Self::Timestamp(e)
    }
}

impl From<DkimCanonicalizationError> for DkimTagValueError {
    fn from(e: DkimCanonicalizationError) -> Self {
        Self::Canonicalization(e)
    }
}

impl From<DkimAlgorithmError> for DkimTagValueError {
    fn from(e: DkimAlgorithmError) -> Self {
        Self::Algorithm(e)
    }
}

impl From<DkimVersionError> for DkimTagValueError {
    fn from(e: DkimVersionError) -> Self {
        Self::Version(e)
    }
}
