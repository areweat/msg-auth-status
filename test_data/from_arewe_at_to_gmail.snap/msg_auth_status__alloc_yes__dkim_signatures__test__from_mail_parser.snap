---
source: src/alloc_yes/dkim_signatures.rs
expression: "&status"
---
DkimSignatures {
    dkim_signatures: [
        DkimSignature {
            v: One,
            a: Rsa_Sha256,
            b: "Q1pcJCMyKT+G69BkGaxzNbuYU1QDacp4K5NkxAYAajJ+l/9eBf\r\n\t5tv+GzwmhWziHknXeKUS0oa1n3PmzP1itsAu53q1zawxYk6rN4Rgtd27jNfHpkrvHmDaZdIl/+j\r\n\tv3KV4bEtFitDJDYQVpx+CyZ382tqWToaqjogBiwZAIrk4RSBeUFb3X3Ymr6Jc8z8VaDnFqK1gxj\r\n\titWw2uGH3NHlQurknRZBkJDZ5jEMq2MzGmCjZ7EaK3xCd8CgPqVnwmJQ7WViOG6f5LxDb8uadJC\r\n\t4E6gH9/CKSUCOmRGFqJLNGSGT1e9hisATRIZQCdsdFYKR5pzlAzHWZhcnFV7lOszvFw==",
            bh: "xer35AYptzY75cgbqHdT+6+\r\n\tJtrP3Ik4+J5hkf9+qRHk=",
            c: Some(
                Relaxed,
            ),
            d: "arewe.at",
            h: "Subject:From:To:Date:Message-ID",
            i: None,
            l: None,
            q: None,
            s: "202405r",
            t: Some(
                Raw(
                    "1717999985",
                ),
            ),
            x: None,
            z: None,
            raw: Some(
                "v=1; a=rsa-sha256; s=202405r; d=arewe.at; c=relaxed/relaxed;\r\n\th=Subject:From:To:Date:Message-ID; t=1717999985; bh=xer35AYptzY75cgbqHdT+6+\r\n\tJtrP3Ik4+J5hkf9+qRHk=; b=Q1pcJCMyKT+G69BkGaxzNbuYU1QDacp4K5NkxAYAajJ+l/9eBf\r\n\t5tv+GzwmhWziHknXeKUS0oa1n3PmzP1itsAu53q1zawxYk6rN4Rgtd27jNfHpkrvHmDaZdIl/+j\r\n\tv3KV4bEtFitDJDYQVpx+CyZ382tqWToaqjogBiwZAIrk4RSBeUFb3X3Ymr6Jc8z8VaDnFqK1gxj\r\n\titWw2uGH3NHlQurknRZBkJDZ5jEMq2MzGmCjZ7EaK3xCd8CgPqVnwmJQ7WViOG6f5LxDb8uadJC\r\n\t4E6gH9/CKSUCOmRGFqJLNGSGT1e9hisATRIZQCdsdFYKR5pzlAzHWZhcnFV7lOszvFw==;",
            ),
        },
        DkimSignature {
            v: One,
            a: Ed25519_Sha256,
            b: "aaPOfWvsZUfQcQrW+iLUckTYeWuJGa3Gi1Lq/tI93eg2eKWSpk\r\n\tZSxeUD5Xq7xnLS5aYMgl7QEsN9rJlQ5BYrBA==",
            bh: "xer35AYptzY75cgbqHdT+6+\r\n\tJtrP3Ik4+J5hkf9+qRHk=",
            c: Some(
                Relaxed,
            ),
            d: "arewe.at",
            h: "Subject:From:To:Date:Message-ID",
            i: None,
            l: None,
            q: None,
            s: "202405e",
            t: Some(
                Raw(
                    "1717999985",
                ),
            ),
            x: None,
            z: None,
            raw: Some(
                "v=1; a=ed25519-sha256; s=202405e; d=arewe.at; c=relaxed/relaxed;\r\n\th=Subject:From:To:Date:Message-ID; t=1717999985; bh=xer35AYptzY75cgbqHdT+6+\r\n\tJtrP3Ik4+J5hkf9+qRHk=; b=aaPOfWvsZUfQcQrW+iLUckTYeWuJGa3Gi1Lq/tI93eg2eKWSpk\r\n\tZSxeUD5Xq7xnLS5aYMgl7QEsN9rJlQ5BYrBA==;",
            ),
        },
    ],
}
